### PR TITLE
`Development`: Replace deprecated before unload event with pending changes guard

### PR DESCRIPTION
**Pull request by coolchock**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Client
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).

### Description

The `beforeunload` event, which is deprecated, was still in use in several parts of the codebase. This pull request replaces the deprecated event with the `PendingChangesGuard` approach, ensuring a more robust and Angular-friendly solution for handling unsaved changes.

Additionally, the `beforeunload` event has been completely removed from the `code-editor-container.component.ts`. Since file saving is now automatically triggered whenever the file content changes, there is no longer a need for the event in this context.

### Steps for Testing

Prerequisites:
- 1 Admin
- 1 Test Server with admin account

The objective is to verify that the warning event, which alerts users about unsaved changes before leaving the page, still functions correctly in all places where the deprecated beforeunload event was replaced.

**standardized-competency-management.component.ts**
1. Go to Server Administration -> Standardized Competencies
2. Expand a category and select a competency
3. Click edit and try to leave the page

**generate-competencies.component.ts**
1. https://artemis-test3.artemis.cit.tum.de/course-management/15/competency-management/generate
2. Enter a course description, generate competencies
3. When competencies were generated, try to leave the page

**course-import-standardized-course-competencies.component.ts**
1. Go to Course Management -> Competency Management
2. Click Import Competencies -> Import Standardized Competencies
3. Select some competencies to import and try to leave the page

**import-course-competencies.component.ts**
1. Go to Course Management -> Competency Management
2. Click Import Competencies -> Import from other courses
3. Select some competencies to import and try to leave the page

**exam-participation.component.ts**
1. Participate in an exam
2. Try to leave the page

**modeling-submission.component.ts**
1. Participate in a modeling exercise
2. Make some changes
3. Try to leave the page

**quiz-exercise-update.component.ts**
1. Try to create/update a quiz exercise
2. Try to leave the page

**text-editor.component.ts**
1. Participate in a text exercise
2. Make some changes
3. Try to leave the page

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

## Summary by CodeRabbit

- **New Features**
	- Introduced `PendingChangesGuard` to prevent navigation away from certain components if there are unsaved changes.
	- Added similar guard for the `TextEditorComponent` to enhance user experience.

- **Bug Fixes**
	- Removed the `@HostListener` decorator and associated unsaved changes warning, simplifying component lifecycle management.

- **Tests**
	- Updated test suites to directly assert the `canDeactivate` method behavior, improving clarity and maintainability.
	- Enhanced tests for `QuizExerciseUpdateComponent` to validate quiz exercise states and error handling.

### CODE REVIEW

**Review by HawKhiem — DISMISSED** [review]
I tested on ts3. All warning events work as expected!

**Review by vinceclifford — DISMISSED** [review]
Tested on TS3, works as expected.

**Review by SimonEntholzer — DISMISSED** [review]
Tested in testing session, all places work still as expected, with the alerts showing up.

**Review by DominikRemo — DISMISSED** [review]
Tested in testing session on ts3. Works well 🚀

**Review by BBesrour — DISMISSED** [review]
Tested in a testing session, works as expected

File: src/test/javascript/spec/component/quiz-exercise/quiz-exercise-update.component.spec.ts [hunk 1]
```diff
@@ -635,16 +635,14 @@ describe('QuizExerciseUpdateComponent', () => {
...
             });
 
             it('should not set event return value to translate if  canDeactivate', () => {
                 comp.pendingChangesCache = false;
-                const ev = { returnValue: undefined };
-                comp.unloadNotification(ev);
-                expect(ev.returnValue).toBeUndefined();
+                const canDeactivate = comp.canDeactivate();
+                expect(canDeactivate).toBeTrue();
             });
```
Comment (pzdr7) [inline src/test/javascript/spec/component/quiz-exercise/quiz-exercise-update.component.spec.ts:646]: Without the HostListener, there's no event here anymore. Can you update the test descriptions? (e.g. should/should not deactivate with pending changes)

File: src/test/javascript/spec/component/competencies/generate-competencies/generate-competencies.component.spec.ts [hunk 2]
```diff
@@ -261,9 +261,8 @@ describe('GenerateCompetenciesComponent', () => {
 
     it('should send a warning when trying to reload', () => {
         generateCompetenciesComponent.isLoading = true;
-        const event = { returnValue: undefined };
-        generateCompetenciesComponent.unloadNotification(event);
-        expect(event.returnValue).toBe(generateCompetenciesComponent.canDeactivateWarning);
+        const canDeactivate = generateCompetenciesComponent.canDeactivate();
+        expect(canDeactivate).toBeFalse();
```
Comment (pzdr7) [inline src/test/javascript/spec/component/competencies/generate-competencies/generate-competencies.component.spec.ts:265]: Consider updating the test description here as well

**Review by pzdr7 — CHANGES_REQUESTED** [review]
Changes look fine. Two small comments on the tests

**Review by pzdr7 — APPROVED** [review]
Code 👍

**Review by BBesrour — APPROVED** [review]
re-approval

**Review by vinceclifford — APPROVED** [review]
Reapproval

**Review by raffifasaro — APPROVED** [review]
code LGTM

**Review by HawKhiem — APPROVED** [review]
Retested on TS3. Works as described

**Pull request merged**